lib/medtronic600/NGPHistoryParser.js

```diff
@@ -618,28 +832,217 @@ class NGPHistoryParser {
+      currentNgpTimestamp.offset));
+
+    for (const event of this.eventsOfTypeReverse(
+        NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_BG_TARGETS,
+        NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_SENSITIVITY,
+        NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_TO_CARB_RATIOS,
+        NGPHistoryEvent.EVENT_TYPE.NEW_BASAL_PATTERN,
+        NGPHistoryEvent.EVENT_TYPE.BASAL_PATTERN_SELECTED)) {
+      const settingsChange = _.cloneDeep(currentSettings);
+
+      switch (event.eventType) {
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_BG_TARGETS:
+          {
+            const targets = event.bgTargets();
+            // Because we read the initial pump settings in mg/dL, always convert settings changes.
+            if (event.units === NGPUtil.NGPConstants.BG_UNITS.MMOL_L) {
+              for (const target of targets) {
+                target.high = Math.ceil(target.high * 18);
+                target.low = Math.ceil(target.low * 18);
+              }
+            }
+            settingsChange.with_bgTarget(targets);
+          }
+          break;
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_SENSITIVITY:
+          {
+            const isfs = event.isfs();
+            // Because we read the initial pump settings in mg/dL, always convert settings changes.
+            if (event.units === NGPUtil.NGPConstants.BG_UNITS.MMOL_L) {
+              for (const isf of isfs) {
+                isf.amount = Math.ceil(isf.amount * 18);
+              }
+            }
+            settingsChange.with_insulinSensitivity(isfs);
+          }
+          break;
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_TO_CARB_RATIOS:
+          {
+            const carbRatios = event.carbRatios();
+            // TODO - Exchange to grams conversion
+            settingsChange.with_carbRatio(carbRatios);
+          }
+          break;
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BASAL_PATTERN:
+          // Basal changes with empty schedules are deletions.
+          if (event.patterns().length === 0 &&
+            _.isObject(settingsChange.basalSchedules[event.patternName])) {
+            delete settingsChange.basalSchedules[event.patternName];
+          }
+          settingsChange.basalSchedules[event.patternName] = event.patterns();
+          break;
+        case NGPHistoryEvent.EVENT_TYPE.BASAL_PATTERN_SELECTED:
+          settingsChange
+            .with_activeSchedule(event.newPatternName);
+          break;
+        default:
+          debug('Unexpected settings change type', event.eventType);
+          throw new Error('Unexpected settings change');
+      }
+      events.push(settingsChange.done());
+    }
+
+    // Now that we're done using currentSettings for cloning, complete it and push to events.
+    events.push(currentSettings.done());
+
+    return this;
+  }
+
+  buildAlarmRecords(events) {
```

Comment:
The pump definitely generates a suspend basal event (I've checked that it does this on a set change, and on a battery swap). But I do need to reverse engineer the alarm message so that we can send the events to the backend.